mining.go

```diff
@@ -2053,3 +2033,374 @@ func (g *BlkTmplGenerator) UpdateBlockTime(header *wire.BlockHeader) error {
+	if g.regenScheduler != nil {
+		if !g.regenScheduler.Reset(dt) {
+			minrLog.Errorf("failed to reset the template regen scheduler")
+			return
+		}
+	}
+
+	g.regenScheduler = time.AfterFunc(dt, func() {
+		go g.regenTemplate()
+		g.cancelScheduledRegen()
+	})
+}
+
+// cancelScheduledRegen terminates a scheduled template regeneration.
+func (g *BgBlkTmplGenerator) cancelScheduledRegen() {
+	if g.regenScheduler != nil {
+		g.regenScheduler.Stop()
+		g.regenScheduler = nil
+	}
+}
+
+// notifySubscribersHandler updates subscribers of newly created block
+// templates. All subscribers are unsubscribed after being updated and
+// required to resubscribe after a template update.
+func (g *BgBlkTmplGenerator) notifySubscribersHandler(ctx context.Context) {
+	minrLog.Debug("Starting notify subscribers handler.")
+	for {
+		select {
+		case t := <-g.notifyChan:
+			g.subscriptionMtx.Lock()
+			for c := range g.subscribers {
+				go func() {
+					c <- t
+					close(c)
+				}()
+
+				delete(g.subscribers, c)
+			}
+			g.subscriptionMtx.Unlock()
+		case <-ctx.Done():
+			minrLog.Debug("Notify Subscribers handler done.")
+			return
+		default:
+			// Non-blocking receive fallthrough
+		}
+	}
+}
+
+// RequestTemplateUpdate subscribes a client for a block template update.  The
+// client is unsubscribed after being updated and required to resubscribe
+// for the next block template update.
+func (g *BgBlkTmplGenerator) RequestTemplateUpdate() chan *BlockTemplate {
+	updateChan := make(chan *BlockTemplate, 1)
+	g.subscriptionMtx.Lock()
+	g.subscribers[updateChan] = struct{}{}
+	g.subscriptionMtx.Unlock()
+	return updateChan
+}
+
+// regenTemplate regenerates the block template.  This must be run as a
+// goroutine.
+func (g *BgBlkTmplGenerator) regenTemplate() {
+	// NOTE: If the blockchain package is updated to provide more
+	// fine grained locking, it will be necessary to add a call to
+	// WaitForChain here.
+
+	// Do not generate block templates if the chain is reorganizing.
+	g.chainReorgMtx.Lock()
+	if g.chainReorg {
+		g.chainReorgMtx.Unlock()
+		return
+	}
+	g.chainReorgMtx.Unlock()
+
+	// Regenerate block templates on mainnet only when the chain is synced.
+	if !g.tg.chain.IsCurrent() && !g.permitConnectionlessMining {
+		return
+	}
+
+	// Pick a mining address at random.
+	rand.Seed(time.Now().UnixNano())
+	payToAddr := g.miningAddrs[rand.Intn(len(g.miningAddrs))]
+
+	// Regenerate the block template.
+	template, err := g.tg.NewBlockTemplate(payToAddr)
+	if err != nil {
+		minrLog.Debugf("block template generation failed: %v", err)
+		return
+	}
+
+	msgBlock := *template.Block
+
+	// In order to efficiently store the variations of block templates that
+	// have been provided to callers, save a pointer to the block keyed by
+	// the merkle root + stake root.  This along with the data that is included
+	// in a work submission is used to rebuild the block before checking the
+	// submitted solution.
+	var merkleRootPair [merkleRootPairSize]byte
```

Comment:
This is not enough to uniquely identify each template in the pool. Observe:

```
2019-03-29 18:35:53.318 [DBG] MINR: Added template for block 07ec5fd6a3283446068f7af0d23cfecd9e72c0fdf2bfc7fbcbd4d4537fe07b8e (height 1) with merkle root pair b0568a679c6b4636d82bed952d9d295d2495cd531890234cf1a3d08885ec9ee90000000000000000000000000000000000000000000000000000000000000000 to pool (size: 1)
2019-03-29 18:35:53.318 [INF] MINR: Received block template building on 5bec7567af40504e0994db3b573c186fffcc4edefe096ff2e58d00523bd7e8a6 (height 0) (1 transactions, 0 stake transactions, 0 votes, -1 in fees, 6 signature operations, 371 bytes, target difficulty 7fffff0000000000000000000000000000000000000000000000000000000000, stake difficulty 0.0002)
2019-03-29 18:36:54.319 [DBG] MINR: Added template for block 7b02462c5ec6b79cb424e1075988dac5503f458f25220a5082fdd9ff02226544 (height 1) with merkle root pair b0568a679c6b4636d82bed952d9d295d2495cd531890234cf1a3d08885ec9ee90000000000000000000000000000000000000000000000000000000000000000 to pool (size: 1)
2019-03-29 18:36:54.319 [INF] MINR: Received block template building on 5bec7567af40504e0994db3b573c186fffcc4edefe096ff2e58d00523bd7e8a6 (height 0) (1 transactions, 0 stake transactions, 0 votes, -1 in fees, 6 signature operations, 371 bytes, target difficulty 7fffff0000000000000000000000000000000000000000000000000000000000, stake difficulty 0.0002)
2019-03-29 18:37:55.318 [DBG] MINR: Added template for block 29634291d71394a4e26915b66126389824e77e5e7d4da9f48b7943c98523c57f (height 1) with merkle root pair b0568a679c6b4636d82bed952d9d295d2495cd531890234cf1a3d08885ec9ee90000000000000000000000000000000000000000000000000000000000000000 to pool (size: 1)
2019-03-29 18:37:55.318 [INF] MINR: Received block template building on 5bec7567af40504e0994db3b573c186fffcc4edefe096ff2e58d00523bd7e8a6 (height 0) (1 transactions, 0 stake transactions, 0 votes, -1 in fees, 6 signature operations, 371 bytes, target difficulty 7fffff0000000000000000000000000000000000000000000000000000000000, stake difficulty 0.0002)
2019-03-29 18:38:56.318 [DBG] MINR: Added template for block 5970b9b7b4548798f3e3d7fc3568ec693d73323b3e2188dd2eb00ca1d364f096 (height 1) with merkle root pair b0568a679c6b4636d82bed952d9d295d2495cd531890234cf1a3d08885ec9ee90000000000000000000000000000000000000000000000000000000000000000 to pool (size: 1)
2019-03-29 18:38:56.318 [INF] MINR: Received block template building on 5bec7567af40504e0994db3b573c186fffcc4edefe096ff2e58d00523bd7e8a6 (height 0) (1 transactions, 0 stake transactions, 0 votes, -1 in fees, 6 signature operations, 371 bytes, target difficulty 7fffff0000000000000000000000000000000000000000000000000000000000, stake difficulty 0.0002)
2019-03-29 18:39:57.319 [DBG] MINR: Added template for block b927770a529d3cdfafba6a7df410293684b6809748cea229f2f906fa161e1bee (height 1) with merkle root pair b0568a679c6b4636d82bed952d9d295d2495cd531890234cf1a3d08885ec9ee90000000000000000000000000000000000000000000000000000000000000000 to pool (size: 1)
2019-03-29 18:39:57.319 [INF] MINR: Received block template building on 5bec7567af40504e0994db3b573c186fffcc4edefe096ff2e58d00523bd7e8a6 (height 0) (1 transactions, 0 stake transactions, 0 votes, -1 in fees, 6 signature operations, 371 bytes, target difficulty 7fffff0000000000000000000000000000000000000000000000000000000000, stake difficulty 0.0002)
```

Notice the the size of the template pool is not increasing because the `merkleRootPair` is not unique. This is because there are no new transactions in between them and thus the merkle roots don't change.  It needs to include the timestamp as well or it would cause miners who submitted a solution to a previous template to get an error that the template can't be found.